src/internal.c

```diff
@@ -222,6 +225,178 @@ static int SSL_hmac(WOLFSSL* ssl, byte* digest, const byte* in, word32 sz,
     int tsip_generatePremasterSecret();
     int tsip_generateEncryptPreMasterSecret(WOLFSSL *ssl, byte *out, word32 *outSz);
 #endif
+#if defined(OPENSSL_EXTRA) && defined(HAVE_SECRET_CALLBACK)
+
+    static int  SessionSecret_callback(WOLFSSL* ssl, void* secret,
+                                                  int* secretSz, void* ctx);
+    static int  SessionSecret_callback_Tls13(WOLFSSL* ssl, int id,
+                       const unsigned char* secret, int secretSz, void* ctx);
+
+    /*
+     * This API is an implementation of SessionSecretCB to be passed
+     * to wolfSSL_set_session_secret_cb()
+     *
+     */
+    static int SessionSecret_callback(WOLFSSL* ssl, void* secret,
+                    int* secretSz, void* ctx)
+    {
+        wolfSSL_CTX_keylog_cb_func logCb = NULL;
+        int msSz;
+        int hasVal;
+        int i;
+        const char* label = "CLIENT_RANDOM";
+        int labelSz = sizeof("CLIENT_RANDOM");
+        int buffSz;
+        byte* log = NULL;
+        word32 outSz;
+
+        (void)ctx;
+
+        if (ssl == NULL || secret == NULL || *secretSz == 0)
+            return BAD_FUNC_ARG;
+        if (ssl->arrays == NULL || ssl->arrays->clientRandom == NULL)
+            return BAD_FUNC_ARG;
+
+        /* get the user-callback func from CTX*/
+        logCb = ssl->ctx->keyLogCb;
+        if (logCb == NULL)
+            return 0;
+
+        /* need to make sure the given master-secret has a meaningful value */
+        msSz   = *secretSz;
+        hasVal = 0;
+        for (i = 0; i < msSz; i++) {
+            if (*((byte*)secret) != 0) {
+                hasVal = 1;
+                break;
+            }
+        }
+        if (hasVal == 0)
+            return 0; /* master-secret looks invalid */
+
+        /* build up a hex-decoded keylog string
+           "CLIENT_RANDOM <hex-encoded client random> <hex-encoded master-secret>"
+           note that each keylog string does not have LF.
+        */
+        buffSz  = labelSz + (RAN_LEN * 2) + 1 + ((*secretSz) * 2) + 1;
+        log     = XMALLOC(buffSz, ssl->heap, DYNAMIC_TYPE_SECRET);
+
+        if (log == NULL)
+            return MEMORY_E;
+
+        XMEMSET(log, 0, buffSz);
+        XMEMCPY(log, label, labelSz -1);     /* put label w/o terminator */
+        XMEMSET(log + labelSz - 1, ' ', 1);               /* '\0' -> ' ' */
+
+        outSz = buffSz - labelSz;
+        Base16_Encode(ssl->arrays->clientRandom, RAN_LEN,
```

Comment:
Need to check return code of Base16_Encode()

Reply:
Added return code check.